src/backend/InvenTree/report/models.py

```diff
@@ -120,531 +143,361 @@ def save(self, *args, **kwargs):
-            )
-        except Exception:
-            page_size_default = 'A4'
-
-        page_size = self.page_size or page_size_default
-
-        if self.landscape:
-            page_size = page_size + ' landscape'
-
-        return page_size
-
-    def context(self, request):
-        """All context to be passed to the renderer."""
-        # Generate custom context data based on the particular report subclass
-        context = self.get_context_data(request)
-
-        context['base_url'] = get_base_url(request=request)
-        context['date'] = InvenTree.helpers.current_date()
-        context['datetime'] = InvenTree.helpers.current_time()
-        context['page_size'] = self.get_report_size()
-        context['report_template'] = self
-        context['report_description'] = self.description
-        context['report_name'] = self.name
-        context['report_revision'] = self.revision
-        context['request'] = request
-        context['user'] = request.user
-
-        # Pass the context through to any active reporting plugins
-        plugins = registry.with_mixin('report')
-
-        for plugin in plugins:
-            # Let each plugin add its own context data
-            try:
-                plugin.add_report_context(self, self.object_to_print, request, context)
-            except Exception:
-                InvenTree.exceptions.log_error(
-                    f'plugins.{plugin.slug}.add_report_context'
-                )
-
-        return context
-
-    def generate_filename(self, request, **kwargs):
+    def generate_filename(self, context, **kwargs):
         """Generate a filename for this report."""
         template_string = Template(self.filename_pattern)
 
-        ctx = self.context(request)
-
-        context = Context(ctx)
+        return template_string.render(Context(context))
 
-        return template_string.render(context)
-
-    def render_as_string(self, request, **kwargs):
+    def render_as_string(self, instance, request=None, **kwargs):
         """Render the report to a HTML string.
 
         Useful for debug mode (viewing generated code)
         """
-        return render_to_string(self.template_name, self.context(request), request)
+        context = self.get_context(instance, request, **kwargs)
+
+        return render_to_string(self.template_name, context, request)
 
-    def render(self, request, **kwargs):
+    def render(self, instance, request=None, **kwargs):
         """Render the template to a PDF file.
 
         Uses django-weasyprint plugin to render HTML template against Weasyprint
         """
-        # TODO: Support custom filename generation!
-        # filename = kwargs.get('filename', 'report.pdf')
+        context = self.get_context(instance, request)
 
         # Render HTML template to PDF
-        wp = WeasyprintReportMixin(
+        wp = WeasyprintReport(
             request,
             self.template_name,
             base_url=request.build_absolute_uri('/'),
```

Comment:
Does this really work if request is None? And do we really need the request object at all in here?

Reply:
@wolflu05 I'll have to look into this further but it does look like we can remove the request object entirely.